src/csharp/Microsoft.Spark/Broadcast.cs

```diff
@@ -170,21 +171,50 @@ private JvmObjectReference CreateBroadcast(SparkContext sc, T value)
             bool encryptionEnabled = bool.Parse(
                 sc.GetConf().Get("spark.io.encryption.enabled", "false"));
 
+            var _pythonBroadcast = (JvmObjectReference)javaSparkContext.Jvm.CallStaticJavaMethod(
+                "org.apache.spark.api.python.PythonRDD",
+                "setupBroadcast",
+                _path);
+
             if (encryptionEnabled)
             {
-                throw new NotImplementedException("Broadcast encryption is not supported yet.");
+                var pair = (JvmObjectReference[])_pythonBroadcast.Invoke("setupEncryptionServer");
+
+                using ISocketWrapper socket = SocketFactory.CreateSocket();
+                socket.Connect(
+                    IPAddress.Loopback,
+                    (int)pair[0].Invoke("intValue"),
+                    (string)pair[1].Invoke("toString"));
+                WriteEncrypted(value, socket.OutputStream);
+                socket.OutputStream.Close();
+                _pythonBroadcast.Invoke("waitTillDataReceived");
             }
             else
             {
                 WriteToFile(value);
             }
 
-            var pythonBroadcast = (JvmObjectReference)javaSparkContext.Jvm.CallStaticJavaMethod(
-                "org.apache.spark.api.python.PythonRDD",
-                "setupBroadcast",
-                _path);
+            return (JvmObjectReference)javaSparkContext.Invoke("broadcast", _pythonBroadcast);
+        }
 
-            return (JvmObjectReference)javaSparkContext.Invoke("broadcast", pythonBroadcast);
+        /// TODO: This is not performant as it writes to stream only after serializing the whole
+        /// value, instead of serializing and writing in chunks like Python.
+        /// <summary>
+        /// Function to write the broadcast value into the encrypted stream.
+        /// </summary>
+        /// <param name="value">Broadcast value to be written to the stream</param>
+        /// <param name="stream">Stream connecting to encryption server to write value to</param>
+        private void WriteEncrypted(object value, Stream stream)
+        {
+            var formatter = new BinaryFormatter();
+            using var ms = new MemoryStream();
+            formatter.Serialize(ms, value);
+            byte[] valueBytes = ms.ToArray();
+            int bytesRemaining = valueBytes.Length;
+            SerDe.Write(stream, bytesRemaining);
+            SerDe.Write(stream, valueBytes, 0, bytesRemaining);
```

Comment:
Can you check if the following works? (so that we don't copy to intermediate bytes arrays)
```csharp
SerDe.Write(stream, ms.Length);
ms.WriteTo(stream);
```

Reply:
Works, changed.